src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -134,27 +151,33 @@ where
 /// Propagates errors from convert shares
 /// # Panics
 /// Propagates panics from convert shares
-pub async fn convert_all_bits<F, C, S>(
+pub async fn convert_all_bits<'a, F, C, S, I>(
     ctx: &C,
-    locally_converted_bits: &[Vec<BitConversionTriple<S>>],
-) -> Result<Vec<Vec<S>>, Error>
+    locally_converted_bits: I,
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<Vec<Vec<Vec<S>>>, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
-    S: ArithmeticSecretSharing<F>,
+    S: ArithmeticSecretSharing<F> + 'a,
+    I: Iterator<Item = Vec<BitConversionTriple<S>>>,
 {
-    let futures = locally_converted_bits
-        .iter()
-        .enumerate()
-        .map(|(bit_num, one_column)| {
-            convert_bit_list(
-                ctx.narrow(&ModulusConversion(bit_num.try_into().unwrap())),
-                one_column,
-            )
-        })
-        .collect::<Vec<_>>();
-    let converted_shares = try_join_all(futures).await?;
-    Ok(converted_shares)
+    try_join_all(
+        (0..num_bits)
+            .step_by(num_multi_bits.try_into().unwrap())
+            .map(|bit_num| {
+                let last_bit = std::cmp::min(num_multi_bits + bit_num, num_bits) as usize;
+                try_join_all(locally_converted_bits.enumerate().map(|(idx, record)| {
```

Comment:
Doesnt compile saying "cannot move out of `locally_converted_bits`, a captured variable in an `FnMut` closure
move occurs because `locally_converted_bits` has type `I`, which does not implement the `Copy` trait"

Reply:
I pushed a small change to your branch that makes this code compile. basically it did not work because you wanted to consume this iterator multiple times within `map` function and Rust wouldn't allow that. [`by_ref`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.by_ref) lets you exclusively borrow this iterator multiple times and it should work in this case, given that you own `locally_converted_bits`

Reply:
The code compiles with this change but the iterator is empty after first iteration.

Reply:
Made it work